### ISSUE TITLE
Why I don't use Styled Components

### ISSUE THREAD

**Issue by gweax** [styled-components/styled-components#637]
With regard to https://twitter.com/_philpl/status/848042319467229184 I'd like to explain in more than 140 chars why I don't use Styled Components.

## Setup

Currently, I work in a project for a large company with hard restrictions both on the hardware as well as the software. I cannot use my own computer, and on the computer I have to work on, I don't have administrative rights. The IDE installed is IntelliJ (at least a decent IDE), and no other IDE can or will be installed in addition (even though VS Code would just require unpacking a zip and whitelisting the plugins url).

That means I am stuck with what is available for IntelliJ and the command line. This heavily influences my reasoning, however it's not the single cause.

I experimented with Styled Components on a medium complex component. Unfortunately, I didn't save the code, so this is written up from memory.

## Syntax checking

In one case, I made a syntax error, which got unnoticed for a while, because of the missing syntax checking. I think it was a superfluous quotation mark at the end of a line. However, since it was in a property that got vendor-prefixed, the first line was fine, the `-webkit` prefixed line worked. In this case, only Firefox was affected, since it used the unprefixed property.

This is a hard to find bug, and the fact that it worked fine in one browser delayed its discovery even further. That was the first argument towards the decision against Styled Components.

## Syntax highlighting

Had there been syntax highlighting, when typing the superfluous quotation mark I could have noticed, there is no matching quotation mark.

## Code completion

I wanted to set the cursor on a disabled button to that forbidden icon. You know what I mean, but you don't know its name from heart either, do you? Writing CSS in JavaScript, it's just a string without meaning. No code completion meant I had to head over to MDN to see it's called `not-allowed`.

Having to look up property names and possible values or even class names used in the markup makes development much slower.

That was the strongest reason to drop Styled Components. I am much more efficient working in an IDE with proper code completion, so I switched back to CSS (Less).

## Linting

Having a linter is nice, but running it in a separate window instead of in the window I'm coding in means the feedback is delayed. You save the code in the IDE, you head over to see the results in the browser, but forgot to look in the linter window... 

## React-Styleguidist

We make a styleguide for the components we create. The tool of our choice is [React Styleguidist](https://github.com/styleguidist/react-styleguidist). However, by wrapping the component via `styled(comp)` the properties don't show up anymore in the generated documentation.

## Summary

In summary, the missing tooling support makes me much slower, and leads to hard to find (of late to find) bugs. There is tooling support for some Editors which brings syntax highlighting and linting. However, without code completion I think I still wouldn't use it, even if the current tools were available in IntelliJ.

## Advantages of Styled Components I miss with CSS

The possibility of Styled Components to change styles at runtime based on the current properties is something I miss when using CSS separately. I solved it in the traditional way with modifier classes. This works for us, since I don't have to set special values passed in at runtime via the ThemeProvider.

**Comment by philpl** [comment]
@mxstbr I think a couple of people have now been burned by hard-to-spot typos. Maybe we can write a tiny linter / or integrate a linter into the babel plugin?

**Comment by mxstbr** [comment]
We have code completion support in Atom and will have it in Sublime soon. JetBrains choosing not to support is sadly their choice, and there isn't much we can do about it, see #176.

I'm really sorry to hear that that's a showstopper since we've been doing everything humanly possible to get tooling support as good as we can, but JetBrains explicitly chose not to support us at this time and thusly there's nothing we can do about that :confused:

----

I agree we need to do something about typos, maybe @thysultan has an idea?

**Comment by thysultan** [comment]
@mxstbr started work on a linter plugin [https://github.com/thysultan/stylis.js/issues/8](https://github.com/thysultan/stylis.js/issues/8) last month, haven't worked on it much lately so it's not yet done.

**Comment by mxstbr** [comment]
Awesome, that's great to hear we should definitely ship that in dev when it's done!

**Comment by Lehakos** [comment]
Autocomplete does not really suffice

**Comment by mxstbr** [comment]
> Autocomplete does not really suffice 

In what sense? What other tooling do you need? Could you elaborate? :laughing:

**Comment by jony89** [comment]
+1 for `React-Styleguidist` .
Having the same issue with storybook + info addon. 

I think styled-components should rethink this, For e.g, add the style inline the component rather than creating new component.
This can cause issues in the future also, for e.g whenever one will want to ref that component, it's just a coincidence that ref by string is now deprecated so the effort **currently** to get a ref for the component is the same, but it can change in the future.

**Comment by mxstbr** [comment]
> it's just a coincidence that ref by string is now deprecated so the effort currently to get a ref for the component is the same, but it can change in the future.

It'll always stay the same, we just pass it through to the `ref` prop. You shouldn't use a string with `innerRef` either:

```JS
<StyledComp innerRef={comp => this.bla = comp} />
```

**Comment by philpl** [comment]
@jony89 I think we had a separate issue for styleguidist and it turns out that the underlying package that parses the propTypes, etc, can enable support for it: https://github.com/styled-components/styled-components/issues/945

> I think styled-components should rethink this, For e.g, add the style inline the component rather than creating new component.

This is not something we'll ever do, or that any component-based css-in-js library will do for that matter. Stylesheets enable great performance optimisations and increased complexity, and the community won't move back to using inline styles.

I'm not quite sure how you're relating inline styling, refs, and "running over the behaviour of the actual component"? The latter being against the best practices that most of us would put forward.

Wrapping components in styled components is actually something we're discouraging whenever possible. The structure that is recommended is container component -> structural components -> presentation components which goes against this pattern.

The `styled(Comp)` pattern for this matter is something that is necessary for react-styleguidist, but needs to be applied with care, without wrapping a structural or container component whenever possible.

**Comment by jony89** [comment]
Hey @philpl , 

> I'm not quite sure how you're relating inline styling, refs, and "running over the behaviour of the actual component"? The latter being against the best practices that most of us would put forward.

How would you address the `storybook`+`info` addon or `React Styleguidist` problem? or any other similar lib ? will you try to create a solution for each lib?

> Wrapping components in styled components is actually something we're discouraging whenever possible

Well, as u said, it's not always possible and usually `structural components` will have style to define width, height, color, etc of the presentation components.

**Comment by jony89** [comment]
@philpl / @mxstbr , would you consider adding a static prop to the StyledComponent, during dev only that will hold the ref to the WrappedComponent? 

So we will have something like : 

```js
<MyStyledExportedComponent.WrappedComponent />
```

I guess I can export as default `MyStyledExportedComponent` and in addition the `WrappedComponent` but I don't like forcing patterns of production code to be used in dev mode.

**Comment by jony89** [comment]
@philpl & @mxstbr ,

Another major issue I found is while using [react2angular](https://github.com/coatue-oss/react2angular) which should automatically resolve angular attribute into react props is not working due to exposing the `StyledComponent` without the actual propTypes.

I'm using my words carefully here, but this can be solved through few options, on the `react2angular` side but on `styled-components` as well, And I'm thinking, why would we just configure propTypes of the `WrappedComponent` into the `StyledComponent` ? 

In some way, they are the props of the `StyledComponent` as well. 

What do you say ?

**Comment by TazmanianD** [comment]
I came across this while checking out styled components and thought I'd add a note specifically about IntelliJ IDEA/WebStorm as the original author mentioned that. It appears there's a plugin that supports styled components and gets you all the goodness you're missing above with code completion and syntax checking: https://plugins.jetbrains.com/plugin/9997-styled-components.

**Comment by imadbz** [comment]
@gweax buddy you should find a better place to work! cos, most of the issues you raised are already solved for VScode as an example. For integrating with other libraries, well, this is opensource JS, fork the projects, find a good solution and share the goodness with rest of us! (ae: Pull Request you solution). This is what makes this community great in the first place and this is how styled-components or any other of the hundred of libraries you use in your project have been built.
